src/main/java/com/frostwire/jlibtorrent/CircularArray.java

```diff
@@ -0,0 +1,96 @@
+package com.frostwire.jlibtorrent;
+
+/**
+ * Created on 12/21/16.
+ *
+ * @author aldenml
+ * @author gubatron
+ */
+public class CircularArray {
+    private final int[] buffer;
+    private int head;
+    private int end;
+
+    CircularArray(int capacity) {
+        buffer = new int[capacity];
+        head = -1;
+        end = -1;
+    }
+
+    void add(int value) {
+        if (head == -1) {
+            head = end = 0;
+            buffer[end] = value;
+            return;
+        }
+        end = (end + 1) % buffer.length;
+        buffer[end] = value;
+        if (end <= head) {
+            head = (head + 1) % buffer.length;
+        }
+    }
+
+    int get(int i) {
+// maybe...
+//        if (i > size()-1) {
+//            throw new ArrayIndexOutOfBoundsException();
+//        }
+        return buffer[(head + i) % size()];
+    }
+
+    int size() {
+        if (head == -1) {
+            return 0;
+        } else if (head <= end) {
+            return 1 + (end - head);
+        } else {
+            return buffer.length;
+        }
+    }
+
+    /**
+     * Returns up to count elements from the head
+     */
+    public int[] tail(int count) {
+        int bufferSize = size();
+        if (bufferSize == 0) {
+            return new int[0];
+        }
+        int tailLength = Math.min(count, bufferSize);
+        int[] result = new int[tailLength];
+        int pointer = 1 + (end - tailLength);
+        if (pointer < 0) {
+            pointer = (buffer.length + pointer) % bufferSize;
+        }
+        int i = 0;
+        while (i < result.length) {
```

Comment:
I thought about doing it like that, but then I thought it would be trickier selecting the offsets and sizes, this way I iterate just as many times as I need to from the correct starting offset.